test/inject.test.tsx

```diff
@@ -254,37 +258,41 @@ describe("inject based context", () => {
     })
 
     test("support static hoisting, wrappedComponent and ref forwarding", () => {
-        class B extends React.Component {
+        class B extends React.Component<any, any> {
+            static foo
+            static bar
+            testField
+
             render() {
                 this.testField = 1
                 return null
             }
         }
-        B.propTypes = {
+        ;(B as React.ComponentClass).propTypes = {
             x: PropTypes.object
         }
-        B.bla = 17
-        B.bla2 = {}
+        B.foo = 17
+        B.bar = {}
         const C = inject("booh")(B)
         expect(C.wrappedComponent).toBe(B)
-        expect(B.bla).toBe(17)
-        expect(C.bla).toBe(17)
-        expect(C.bla2 === B.bla2).toBeTruthy()
-        expect(Object.keys(C.wrappedComponent.propTypes)).toEqual(["x"])
+        expect(B.foo).toBe(17)
+        expect(C.foo).toBe(17)
+        expect(C.bar === B.bar).toBeTruthy()
+        expect(Object.keys(C.wrappedComponent.propTypes!)).toEqual(["x"])
 
-        const ref = React.createRef()
+        const ref = React.createRef<B>()
 
         render(<C booh={42} ref={ref} />)
-        expect(ref.current.testField).toBe(1)
+        expect(ref.current?.testField).toBe(1)
```

Comment:
Isn't the `?` safer? I don't see the benefit of switching to `!` here.

Reply:
We do not need safety here. `ref.current` cannot be `null` in these cases.

Also this is the answer to https://github.com/mobxjs/mobx-react/pull/826#discussion_r374013294.

Reply:
But it certainly cannot hurt to have this _safety_ here. Later on, someone might modify the test, a value can be _nullable_ and this operator would hide the error.

Reply:
It can hurt because it can hide the error if in the future by mistake ref.current becomes `null`. The non-null assertion operator allows us to clearly express in code our assurance that the expression cannot be null.

Reply:
Wait, what? I think you got that quite opposite :) The non-null assertion operation basically tells TS to shut up and don't warn if the value is possibly nullable. The optional chaining adds runtime code that makes sure it won't be a null...

https://www.typescriptlang.org/play/index.html#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wChS0IA7AZ3iMzgF5FiYA6AVxqWUwB4A3nBhI6AMWBIANgBMAXHDpRgVAOZwAvnAA+cKp2nSAfAAoDRgJTkG7NJyhEqHURKlzSt+46TOAhOyuMJIysp5ImHYOTjAA-IFiwe6yQA

Reply:
I'm not sure it really matters very much.  Either way the test will fail, and that's the important thing.  I think you're both right.

> The non-null assertion operator allows us to clearly express in code our assurance that the expression cannot be null. - vkrol

> The non-null assertion operation basically tells TS to shut up and don't warn if the value is possibly nullable. - FredyC


![image](https://user-images.githubusercontent.com/12650339/73651641-b14e4b80-4639-11ea-93d7-8739e309d204.png)
![image](https://user-images.githubusercontent.com/12650339/73651683-c5924880-4639-11ea-9a38-b1cff0adf81e.png)

Reply:
Yea, you are right, in this context, it doesn't matter that much. The only slight difference is that with optional chaining you can spot an error while coding. That's one of the reasons I love TS that I don't need to wait for the runtime to tell me something is wrong.

Reply:
Fair point.

Reply:
So you're aware, I'm using the non null assertion operator in `disposeOnUnmount.test.tsx` on line 325 and 333 because it was much cleaner than adding a bunch of question marks in every expect.  I'm leaving it as it is, but let me know if you want me to change it.

Reply:
"Cleaner"? :) I don't follow where that comes from. That single character does not look that different.

Reply:
I mean using 

```
const el = ref.current!
expect(el.property).toBe("test")
expect(el.property2).toBe("test2")
```
vs.

```
const el = ref.current
expect(el?.property).toBe("test")
expect(el?.property2).toBe("test2")
```
The next commit will use the latter.  It's more nit-picky than anything.

Reply:
> The non-null assertion operation basically tells TS to shut up and don't warn if the value is possibly nullable. The optional chaining adds runtime code that makes sure it won't be a null...

Yes, and this is the right behavior in principle here, because current cannot be null under any circumstances. And if it suddenly becomes null for some reason, this is an error in our code.

Reply:
When this code was written in JS we did not add if, then why do we add it as optional chaining now?

Reply:
Using optional chaining in this code, we sort of say that it can be `null` and it is normal. Using a non-null assertion operator we say that there can never be `null` and this reflects reality.

Reply:
Let's not nitpick in this, as @ynejati pointed out there is a very little difference except one thing is runtime and other is compile time.

Personally, I try to avoid `!` as much as I can because it can lead to nasty bugs. You are bypassing type checker for some sort of convenience. It was useful before optional chaining was a thing.